### DISCUSSION TITLE
Support for GPS RTK2

### DISCUSSION THREAD

**Discussion by willishf** [iNavFlight/inav#8760] [Ideas]
We have setup GPS RTK2 at our farm property and get 1cm accuracy. For a specific robot application we have two GPS receivers on the robot separated by 3ft/90cm. With the 1cm accuracy we get very high precision on robot heading for navigation. In particular, the robot doesn't need to be moving to determine heading and/or pitch as an example. 

For a large agricultural drone was thinking it would be interesting to test the same setup for an accurate GPS position. Could also extend the ability to calculate heading/pitch/roll/yaw depending on using multiple GPS units on the drone. 

Would need the ability to do the heading/pitch/roll/yaw calculations based on 2 or more GPS locations that are then combined and sent to the iNAV FC software. Refresh rate is a concern but could also serve as a correction factor for drift related to gyro/magnetometer/accelerometer. 

How we are addressing this on the robot use case is we have multiple sources of positional information that can indicate confidence in the accuracy of the measurement and use GPS heading over a compass heading as an example. If GPS goes away for a failure or the GPS determines it has low confidence in its GPS-calculated heading(hovering), then use the compass heading. If the compass/magnetometer heading is changing more than it should(near power lines or high vibration), then use the GPS heading even if it has lower confidence. 

Just some thoughts on possible functionality that can be added in the future to help on autonomous flight when you are trying to determine the accuracy/reliability of sensors.

**Comment by rcjarvis** [comment]
HI Willishf, I have tried with ardupilot and pixhawk for hours to get my high precision rover to work.... can you lead me a little on how to get rtk to work with Inav?  I made  the rover work on normal gps ( Speedy Bee f7v3) on Inav but it lacks the precicion when running in forests and  around buildings where gps signals are compromised... Thanks for considering !   As per your request on integrating rtk on drones. I will do when I get the ground version running... a drone equiped with over 2000Cad worth of gps only is kind of making me nervous  ;)

**Comment by rcjarvis** [comment]
By the way, Inav was set up and running programmed led's and rth gps in less than 45 mins...lol ...

**Comment by willishf** [comment]
Did you get rtk working in an open field? Do you have a base station setup
that you have submitted collected gos data to get your precise location of
the rtk base station?

If you get all that working in an open field then the errors you are seeing
is from lack of direct line if site because of trees. I suspect
rtk won't work in a covered forest.

Using AI is going to be the long term solution for navigation in these
types of environments.

**Comment by rcjarvis** [comment]
HI, and thank you for such a fast response… I could not even turn the motors on motor test, so no I did not get rte working yet.  The gps trials I did were only with the speedy bee and a bn 880 gps unit. And would not get good repeatable patterns ( Building a lawn mower and snow blower rover and the core will be used for  surveying on forest land and waste lands. SO my question is more about integration of rte on Inav because Ardupilot and a pihawk 4, and a 6c mini got me to waste over 72 hours just tying to get th Pihawks to run.. The pixhawk 4 runs but it is erratic… like rc copter tail hunting on gps but not on heading hold ( so pid’s are fine) But the 6c mini keeps getting bad hrs and ‘controller won’t allow motors to turn’ on motor test tab. I got the speedy bee running in no time…But again… precision Mathers in my case.

So if you can share with me what fc you use on Inav and how you injected the gps for RTK… your price is mine !

Daniel Gervais

**Comment by willishf** [comment]
I have attached a poster we did for a conference with some details. No
longer an active project(did it for the conference).

To do GPS RTK requires a special GPS that communicates and can calculate
phase correction or the lie being told so that bad people can't use GPS to
a precise location. You need a Zed F9P GPS as a base station where the
precise location is known and then you need a Zed F9P GPS on the robot. You
then need to communicate from the base station the correction data in real
time so that on the robot it can see what it needs to adjust based on
difference phase data from what is moving vs at a known GPS location. There
are some third party services that have installed enough RTK base stations
that if one is near by you can get the correction data without installing
your own. Mileage will very.

Look up tutorials on Zed F9P at sparkfun and there is a reason it is $250
each.

I haven't been able to find a GPS unit that can do RTK for less than the
sparkfun unit.

I would spend time learning about visual odometry and AI as the way to
solve the problem when you are in spaces where you do not have clear line
of site to the satellites. All the bouncing signals from tall objects know
what to correct for that problem even with RTK.

**Comment by sensei-hacker** [comment]
Be aware, if you're getting multipath from signals bouncing off buildings or similar, RTK won't help you.
RTK would just give you *exactly* the wrong location.

It would more precisely tell you where the wall is, rather than where you are.

**Reply by rcjarvis** [reply]
Hi, I thought so, we had lots of multipathing with video inside steel buldings with video , so gps will also probably be affected

**Comment by rcjarvis** [comment]
Welll thank you very Much, I already purchased the overpriced F9P Survey for base and the just as overpriced Mosaic H for onboard yaw correction. But I have spent over 90 hours tweeking Ardupilot and I can’t stop a ‘aileron’ misbehaviour so I will look into this since I have proven my radio (radiometer tx16S with Hall effects sticks) because all is mighty fine on all my Inav boards.. Thanks a million I will check it out (ZED F9P) And yes, we were planning to install lidars and sonars to prevent impacts of things and livings !

I can not see attached poster !?!??

**Reply by sensei-hacker** [reply]
Also I was just reminded, if you need very high accuracy you might consider UWB beacons such as those from Estimote, if they make sense in your application.

They work well when you can do some setup at the location ahead of time. You could put a beacon in each corner of the property / area of operation. The vehicle then tracks relative to the four corners, rather than relative to satellites far off in space. The difference between measuring your location from the front door rather than from the space station.

**Reply by rcjarvis** [reply]
Interesting, but they only detect each other, not objects or beings ?  So you limit (geo-fence) the terrain with these ... Again, interesting...

**Comment by utrinque96** [comment]
I got RTK module with included 2 antennas supporting UBX protocol for less than $100 at: 
https://pl.aliexpress.com/item/1005006198957261.html?spm=a2g0o.order_list.order_list_main.41.10e91c2499bsiE&gatewayAdapt=glo2pol

It is pretty compact but weight is around 150g (10-13 inch drone would be able to handle this) so for 7 inch drone it seems probably to big to be practical. 
It can work as ordinary GPS with e.g. F405 FC. I tested it with JHEMCU GF30F435 FC. It more pracise than ordinary GPS even without RTK corrections. I assume that only Ardupilot supports NTRIP protocol via telemetry needed to send corrections from Rover or PC. I am wondering if NTRIP data can be sent from PC/mobile/ground station via WiFi to ELRS radio and transferred using e.g. MAVLINK protocol to FC to provide RTK accuracy around 1-2cm? Is ELRS bandwidth good enough? I have GPS coordinates of my quad during flight displayed on my ELRS radio display all the time via telemetry. Can it work the other way? 

Hint:
https://www.youtube.com/watch?v=HTaKspktHhk&t

Nice:
<img width="5241" height="3696" alt="image" src="https://github.com/user-attachments/assets/5ae4c73c-b4aa-4889-a5a6-68710cff954d" />

https://ardupilot.org/copter/docs/common-rtk-correction.html
